pkg/core/mpt/trie.go

```diff
@@ -314,32 +359,69 @@ func makeStorageKey(mptKey []byte) []byte {
 // new node to storage. Normally, flush should be called with every StateRoot persist, i.e.
 // after every block.
 func (t *Trie) Flush() {
-	t.flush(t.root)
+	for h, cnt := range t.updated {
+		if cnt != 0 {
+			node := t.nodes[h]
+			if node == nil {
+				panic("item not in trie")
+			}
+			if t.refcountEnabled {
+				t.updateRefCount(h, cnt, node)
+			} else if cnt > 0 {
+				_ = t.Store.Put(makeStorageKey(h.BytesBE()), node)
+			}
+		}
+	}
+	t.initial = make(map[util.Uint256]int)
+	t.updated = make(map[util.Uint256]int)
+	t.nodes = make(map[util.Uint256][]byte)
```

Comment:
But `initial` and `nodes` data can still be left until compaction happens.

Reply:
Fixed.

Reply:
Removing items from map if they were not changed or removed from the store.

Reply:
You need to drop items from this cache during `Collapse`, otherwise you're effectively leaking memory.

Reply:
What do you mean? The map is updated in `Collapse`.